### ISSUE TITLE
Definitions of uniqueness and existential uniqueness

### ISSUE THREAD

**Issue by benjub** [metamath/set.mm#3485]
As the one-year period for removal of obsolete (*.OLD) statements is approaching for many statements about unique existence and uniqueness, I think we should postpone there deletion since I'm having terrible doubts on whether the definition change I proposed last year (effective 30-Sep-2022) was a good idea. (I still think the splitting into two subsections is better and make things clearer, so that it was a good decision.)

As for the definitions themselves, it seemed more natural to define unique existence as the conjunction of two more elementary notions, existence and uniqueness. That is,
```
df-mo $a |- ( E* x ph <-> E. y A. x ( ph -> x = y ) ) $.
df-eu $a |- ( E! x ph <-> ( E. x ph /\ E* x ph ) ) $.
```
(this is the current situation), compared to the older definitions, which are kept as
```
eu6 $p |- ( E! x ph <-> E. y A. x ( ph <-> x = y ) ) $=
moeu $p |- ( E* x ph <-> ( E. x ph -> E! x ph ) ) $=
```

One problem is that `df-mo` is not correct on empty domains, since there, it is false (since it begins with an existential quantifier), but uniqueness should be considered to tautologically hold on the empty domain.  I wrote a definition which would work,
```
df-bj-mo $a |- ( E** x ph <-> A. z E. y A. x ( ph -> x = y ) ) $.
```
preceded in my mathbox by justification theorems on a very weak axiomatic. But having an extra quantifier may be a bit cumbersome, and I'm not sure yet it would be good to adopt it.  One would need to check the respective strengths of these definitions on that weak axiomatic (basically, no existence axiom, so ax-1--ax-5 with `$a |- A. x E. y y = x $.` (**edit:** and the fact that equality is an equivalence relation, or at least symmetric, as in `bj-cbval`)), and also compare it to the more intuitive
```
$a |- ( E*** x ph <-> A. y A. z ( ( [ y / x ] ph /\ [ z / x ] ph ) -> y = z ) ) $.
```

An associated problem is that the current definitions that I proposed bring maybe more auxiliary axioms (ax-10--ax-13) than necessary.

I apologize for not looking into all the possible consequences of this definition change before proposing it.  Fortunately, all theorems needed to go back and forth between the two definition pairs, and to recover them from each other, are still there, so reverting would be rather easy. I should thank Gérard Lang whose questions on the empty domain problem, which looked small at first, made me realize there might be a bigger problem with the new definitions.

In the short term, we could simply change the "obsolete as of ..." dates of the concerned theorems, e.g., change "2022" to "2023", or "Sep" and "Oct" to "Dec" (there are about 10 to 20 of them) to give us some time to think this true.

What do you think ?

**Comment by tirix** [comment]
Yes, changing the "obsolete as of ..." dates seems to be the most reasonable way to go.

**Comment by wlammen** [comment]
I'm not sure what the empty domain is worth in a context where we build up a theory of sets.  At the point of introduction of E* we long passed ax-6 which guarantees us the existence of an object.  The problem here seems to be that ax-6 may not appear in the axiom list of some early theorems developing uniqueness, which can be seen kind of incorrect.

So why not fix it by
df-mo $a |- ( E* x ph <-> ( E. y y = x -> E. y A. x ( ph -> x = y ) ) )

and its first application is
dfmo $a |- ( E* x ph <-> E. y A. x ( ph -> x = y ) )
proven by ax6ev, a1bi, df-mo and bitri.

Then replace df-mo with dfmo elsewhere.  This way we formally introduce a definition of E* that keeps its intended semantics even in the empty domain, but thereafter we focus ourselves on set theory, ruling out a corner case for which E* is not made anyway.  All proofs carry over unmodified (I hope).

One can still derive on the side alternative definitions directly from df-mo like the ones given in the first comment above.
Some easy results like moimi are likely provable without ax-6 under the regime of the here introduced df-mo as well. 

Or, of course, revert to the old scheme.  This raises other forms of discussions, like "Why does the more complex and specialized E! x precede the more general E*?" .  May also require on the average more axioms and proof lines than the other way round.  One should note in that case the reason in the comment of df-eu.

**Comment by benjub** [comment]
> I'm not sure what the empty domain is worth in a context where we build up a theory of sets.

You're right that it's not important in itself, and as I wrote above it looked to my like a small problem at first when Gérard pointed it out to me. But often, looking at what happens in degenerate cases may shed some light on another inadequacy of a definition. In the present case, the more important potential inadequacy is the use of maybe more auxiliary axioms than necessary (e.g., ax-10--12 in eu6).

> So why not fix it by df-mo $a |- ( E* x ph <-> ( E. y y = x -> E. y A. x ( ph -> x = y ) ) )
> 
> and its first application is dfmo $a |- ( E* x ph <-> E. y A. x ( ph -> x = y ) ) proven by ax6ev, a1bi, df-mo and bitri.
> 
> Then replace df-mo by dfmo elsewhere. This way we formally introduce a definition of E* that keeps its intended semantics even in the empty domain, but thereafter we focus ourselves on set theory, ruling out a corner case for which E* is not made anyway. All proofs carry over unmodified (I hope).

I think it's worth considering your proposal indeed. It solves the problem of the empty domain, but the question about auxiliary axioms remains the same.

> One can still derive on the side alternative definitions directly from df-mo like the ones given in the first comment above. Some easy results like moimi are likely provable without ax-6 using the new df-mo as well.
> 
> Or, of course, revert to the old scheme. This raises other forms of discussions, like "Why does the more complex and specialized E! x precede the more general E*?" . May require on the average more axioms and proof lines than the other way round. One should note in that case the reason in the comment of df-eu.

Indeed, the appeal of defining existential uniqueness as the conjunction of two more elementary notions was what prompted me to propose the current definition last year. Maybe it should stay like that, or maybe revert, or use your proposal above, or use `E**`, or `E***`. I am not sure, and this is why I proposed to delay for some time the deletions of OLD theorems to give us some time to ponder these proposals.

**Comment by wlammen** [comment]
`...the question about auxiliary axioms remains the same`
Following my suggestion you do not add extra ax-10 - ax-13 axioms to any proof in the current version. Since the transition from df-mo to dfmo (which is the current definition) uses ax-mp to ax-3 and ax-6 only, you will see (very) occasionally an extra ax-6 in the axioms list of theorems based on E* that wasn't there before.  In the presence of ax-6 the problem Gérard Lang pointed to does not exist any longer.  That's the idea behind my suggestion.

One can elaborate on the critique of Gérard Lang a bit further, and ask under what circumstances does the expression E. y A. x ( ph -> x = y ) at least remotely convey the idea of uniqueness.  We have seen that an empty universe (ax-6 not valid)  is not sufficient.  Can we drop ax-5?  Difficult, given that a dv condition on ph and y must be fulfilled.  It is introduced to assert that a variation of y will not affect the ph under investigation.  Can equality be somehow weird (ax-7 invalid), i.e. not simultaneously transitive, reflexive, symmetric?  If for example one special set is equal to all others, all other combinations are unequal, then E* x T. seems to hold, contrary to expectations.  It might be sensible to only define E* in universes where a minimum of sanity is given.  The expression x = x may be picked to achieve this goal.  equid is proven from all axioms ax-mp up to ax-7.
df-mo $a |- ( E* x ph <-> ( x = x /\ E. y A. x ( ph -> x = y ) ) )

The latter paragraph boils down to whether you see E* as an abbreviation for a frequent formula popping up anywhere (even in an empty universe!), no underlying meaning automatically assumed, or whether it is meant as a suggestive shortcut conveying an idea.  A similar discussion is made in  #3198 .

**Comment by icecream17** [comment]
~ dfmo : mp to 7 <ins>and 10</ins> and 12 (edited)
~ dfeu : mp to <del>5</del><ins>4</ins> (edited)

~ moeu : mp to 6
~ eu6 : mp to 7 and 10 to 12

It's interesting that (step 23 of dfmo) is provable with mp to 7 and 12
![image](https://github.com/metamath/set.mm/assets/58114641/cca3fc92-290d-4d36-8420-2487dc2af96b)
But exsb (the reason why eu6 uses ax-11)... uses ax-11 (and ax-12)
![image](https://github.com/metamath/set.mm/assets/58114641/94d44143-5dfa-4eda-a692-b3974c6ba7ed)

in fact...

```
eu3v
E! x ph <-> (E. x ph /\ E. y A. x ( ph -> x = y ) )

dfmo step 25 commuted
E. y A. x ( ph -> x = y ) <-> (E. x ph -> E. y A. x ( ph <-> x = y ))

E! x ph <-> (E. x ph /\ ( E. x ph -> E. y A. x ( ph <-> x = y )))

E! x ph <-> (E. x ph /\ E. y A. x ( ph <-> x = y ) )

exsbim et al
E. y A. x ( ph <-> x = y ) -> E. x ph

E! x ph <-> E. y A. x ( ph <-> x = y )
```
which proves eu6 without ax-11

So now
~ dfmo : mp to 7 and 10 and 12
~ dfeu : mp to 4

~ moeu : mp to 6
~ eu6 : mp to 7 and 10 and 12

**Comment by wlammen** [comment]
Nice drill down!

`It's interesting that (step 23 of dfmo) is provable with mp to 7 and 12.`
ax-10 is needed by nfia1 (step 13 of dfmo)
So it is actually mp to 7 and 10 and 12

And ~ dfeu : mp to 3

**Comment by wlammen** [comment]
One should revise exsb and kick out ax-11 on the next occasion, and independently of this issue.  I can do it, but I feel better if @icecream17 is ready for this task, as he deserves the credits.

**Comment by wlammen** [comment]
Uniqueness vs "At most one".  Uniqueness is (currently) based on equality.  If sets for which an expression ph is true are all equal to a given one, then uniqueness is assumed.  To simplify matters here, let equality be an equivalence relation, so it defines equivalence classes of sets.  This still allows for multiple sets per such class distinguishable by other means.  ax12v is needed to rule this out: "equality is the finest grained logical expression".

**Comment by benjub** [comment]
I postponed the 'obsolete' dates for eight df-mo/df-eu-related theorems from Oct-2022 to 31-Dec-2022 in #3511. This gives us around three months to think the options through. We can discuss them here, so I do not close this issue.

Side remark: @wlammen, can you refrain from making too many edits to posts here: it makes it harder to follow the threads (also, the email notifications we get only mention the original version, and not the successive edits, at least with my settings).

**Comment by jkingdon** [comment]
My apologies if this has already been considered and/or would increase axiom usage, but have we thought about making the `E*` definition some variant of https://us.metamath.org/mpeuni/mo4.html ?

Granted, my brain is a bit influenced by iset.mm in which many things are different (especially related to `E*`, less so `E!`).

**Comment by benjub** [comment]
> My apologies if this has already been considered and/or would increase axiom usage, but have we thought about making the `E*` definition some variant of https://us.metamath.org/mpeuni/mo4.html ?

Up to replacing implicit substitution with explicit substitution, this is similar to `E***` above (but `E***` is more "symmetric" than mo3 or mo4). Or you were thinking of something else ?  Note that I called that definition "intuitive" above ;)

> Granted, my brain is a bit influenced by iset.mm in which many things are different (especially related to `E*`, less so `E!`).

That's a beneficial influence !  I should have said above that a definition which is correct in intuitionistic logic should be preferred.  Indeed, the old definition of `E*` is weaker intuitionistically than the current one.

**Comment by wlammen** [comment]
`can you refrain from making too many edits to posts here`
Sorry for any inconvenience.  As long as the contents is not essentially touched (fixing glitches, typos, adding missing links, clarifications etc) I'd like to still edit my posts.
This one https://github.com/metamath/set.mm/issues/3485#issuecomment-1722835092 was indeed heavily edited.  It was because the first idea turned out to be too vague and was developed into its final shape over a few hours, while I was still working.  Every one to two hours I interrupted my work for a coffee break, and updated the post to the then state of knowledge.  I cannot afford this kind of interleaved work normally, so I hope this is a rare event.

**Comment by benjub** [comment]
> Sorry for any inconvenience. As long as the contents is not essentially touched (fixing glitches, typos, adding missing links, clarifications etc) I'd like to still edit my posts. This one [#3485 (comment)](https://github.com/metamath/set.mm/issues/3485#issuecomment-1722835092) was indeed heavily edited. It was because the first idea turned out to be too vague and was developed into its final shape over a few hours, while I was still working.

No worries, and indeed typos and minor edits are fine. It gets a bit harder with heavier edits indeed.

> Every one to two hours I interrupted my work for a coffee break

too much coffee !

**Comment by jkingdon** [comment]
> > My apologies if this has already been considered and/or would increase axiom usage, but have we thought about making the `E*` definition some variant of https://us.metamath.org/mpeuni/mo4.html ?
> 
> Up to replacing implicit substitution with explicit substitution, this is similar to `E***` above

Indeed it is (I had missed `E***` as evidently I was skimming more than I realized).

> > Granted, my brain is a bit influenced by iset.mm in which many things are different (especially related to `E*`, less so `E!`).
> 
> That's a beneficial influence ! I should have said above that a definition which is correct in intuitionistic logic should be preferred. Indeed, the old definition of `E*` is weaker intuitionistically than the current one.

I suspect it would be necessary to try proving things in iset.mm under various definitions to get a lot of insight into this (copy-paste from set.mm to iset.mm might be a good technique to find out which proofs are using missing theorems). I have been operating under the assumption that many of the `E*` theorems which fail due so for unavoidable reasons (like https://us.metamath.org/mpeuni/snex.html which at least feels related). But I don't know that for sure, with either iset.mm's current definitions or other ones.

**Comment by jkingdon** [comment]
> heavily edited. It was because the first idea turned out to be too vague and was developed into its final shape over a few hours

It isn't an especially big deal (at least not to me) one way or the other, but in such circumstances at least consider using the strikethrough (~~like this~~) for the vague comment which allows updating something while perhaps being slightly less confusing. Sometimes people also use language like `UPDATED`, which also is a way of providing that sort of signal.
